pkg/webhook/preflight/preflight.go

```diff
@@ -0,0 +1,156 @@
+	Allowed bool
+	Field   string
+	Message string
+	Warning string
+	Error   bool
+}
+
+type Check = func(ctx context.Context) CheckResult
+
+type Checker interface {
+	// Init decides which of its checks should run for the cluster. It then initializes the checks
+	// with common dependencies, such as an infrastructure client. Finally, it returns the initialized checks,
+	// ready to be run.
+	//
+	// Init should not store the context `ctx`, because each check will accept its own context.
+	// Checks may use both the client and the cluster.
+	Init(ctx context.Context, client ctrlclient.Client, cluster *clusterv1.Cluster) ([]Check, error)
+}
+
+type WebhookHandler struct {
+	client   ctrlclient.Client
+	decoder  admission.Decoder
+	checkers []Checker
+}
+
+func New(client ctrlclient.Client, decoder admission.Decoder, checkers ...Checker) *WebhookHandler {
+	h := &WebhookHandler{
+		client:   client,
+		decoder:  decoder,
+		checkers: checkers,
+	}
+	return h
+}
+
+func (h *WebhookHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation == admissionv1.Delete {
+		return admission.Allowed("")
+	}
+
+	cluster := &clusterv1.Cluster{}
+	err := h.decoder.Decode(req, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Checks run only for ClusterClass-based clusters.
+	if cluster.Spec.Topology == nil {
+		return admission.Allowed("")
+	}
+
+	resp := admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{
+			Allowed: true,
+			Result: &metav1.Status{
+				Details: &metav1.StatusDetails{},
+			},
+		},
+	}
+
+	// Initialize checkers in parallel.
+	type ChecksResult struct {
+		checks []Check
+		err    error
+	}
+	checksResultCh := make(chan ChecksResult, len(h.checkers))
+
+	wg := &sync.WaitGroup{}
+	for _, checker := range h.checkers {
+		wg.Add(1)
+		result := ChecksResult{}
+		result.checks, result.err = checker.Init(ctx, h.client, cluster)
+		checksResultCh <- result
+		wg.Done()
+	}
+	wg.Wait()
+	close(checksResultCh)
+
+	// Collect all checks.
+	checks := make([]Check, 0)
+	for checksResult := range checksResultCh {
+		if checksResult.err != nil {
+			resp.Allowed = false
+			resp.Result.Code = http.StatusInternalServerError
+			resp.Result.Message = "failed to initialize preflight checks"
+			resp.Result.Details.Causes = append(resp.Result.Details.Causes, metav1.StatusCause{
+				Type:    metav1.CauseTypeInternal,
+				Message: checksResult.err.Error(),
+				Field:   "", // This concerns the whole cluster.
+			})
+			continue
+		}
+		checks = append(checks, checksResult.checks...)
+	}
+
+	// Run all checks in parallel.
+	resultCh := make(chan CheckResult, len(checks))
+	for _, check := range checks {
+		wg.Add(1)
+		go func(ctx context.Context, check Check) {
+			result := check(ctx)
+			resultCh <- result
+			wg.Done()
+		}(ctx, check)
+	}
+	wg.Wait()
+	close(resultCh)
+
+	// Collect check results.
+	for result := range resultCh {
```

Comment:
Very clean, on any Error to run a Check or when a check returns a `Allowed==false` the `resp` will always be set to `resp.Allowed = false` and will keep appending to the causes.

(just mainly thinking out loud)

Reply:
That's right. We want to keep track of whether we had a rejection, or an error, but collect all the check results, regardless.